File: sdk/java-sdk-spring/src/main/java/kalix/javasdk/impl/client/ComponentClientImpl.java
/*
 * Copyright (C) 2021-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package kalix.javasdk.impl.client;

import kalix.javasdk.Metadata;
import kalix.javasdk.client.ActionClient;
import kalix.javasdk.client.ComponentClient;
import kalix.javasdk.client.EventSourcedEntityClient;
import kalix.javasdk.client.ValueEntityClient;
import kalix.javasdk.client.ViewClient;
import kalix.javasdk.client.WorkflowClient;
import kalix.spring.impl.KalixClient;

import java.util.Optional;

public class ComponentClientImpl implements ComponentClient {

  private final KalixClient kalixClient;

  private Optional<Metadata> callMetadata = Optional.empty();

  public ComponentClientImpl(KalixClient kalixClient) {
    this.kalixClient = kalixClient;
  }

  public void setCallMetadata(Metadata callMetadata) {
    this.callMetadata = Optional.of(callMetadata);
  }

  public void clearCallMetadata() {
    this.callMetadata = Optional.empty();
  }

  @Override
  public ActionClient forAction() {
    return new ActionClient(kalixClient, callMetadata);
  }


  @Override
  public ValueEntityClient forValueEntity(String valueEntityId) {
    return new ValueEntityClientImpl(kalixClient.runtimeComponentClients().valueEntityClient(), callMetadata, valueEntityId, kalixClient.executionContext());
  }

  @Override
  public EventSourcedEntityClient forEventSourcedEntity(String eventSourcedEntityId) {
    return new EventSourcedEntityClientImpl(kalixClient.runtimeComponentClients().eventSourcedEntityClient(), callMetadata, eventSourcedEntityId, kalixClient.executionContext());
  }


  @Override
  public WorkflowClient forWorkflow(String workflowId) {
    return new WorkflowClientImpl(kalixClient.runtimeComponentClients().workFlowClient(), callMetadata, workflowId, kalixClient.executionContext());
  }


  @Override
  public ViewClient forView() {
    return new ViewClientImpl(kalixClient.runtimeComponentClients().viewClient(), callMetadata, kalixClient.executionContext());
  }

}

File: sdk/java-sdk-spring/src/main/scala/kalix/javasdk/impl/client/ComponentClientImpl.scala
/*
 * Copyright (C) 2021-2024 Lightbend Inc. <https://www.lightbend.com>
 */

package kalix.javasdk.impl.client

import akka.annotation.InternalApi
import kalix.javasdk.Metadata
import kalix.javasdk.client.ComponentClient
import kalix.javasdk.client.EventSourcedEntityClient
import kalix.javasdk.client.ValueEntityClient
import kalix.javasdk.client.WorkflowClient
import kalix.javasdk.client.ActionClient
import kalix.javasdk.client.ViewClient
import kalix.javasdk.spi.{ ComponentClients => RuntimeComponentClients }

import scala.concurrent.ExecutionContext

/**
 * Note: new instance per call since it includes call metadata
 *
 * INTERNAL API
 */
@InternalApi
private[javasdk] final case class ComponentClientImpl(runtimeComponentClients: RuntimeComponentClients)(implicit
    ec: ExecutionContext)
    extends ComponentClient {

  // Volatile since the component client could be accessed in nested/composed futures
  @volatile var callMetadata: Option[Metadata] = None

  override def forAction(): ActionClient = ActionClientImpl(runtimeComponentClients.actionClient, callMetadata)

  override def forValueEntity(valueEntityId: String): ValueEntityClient =
    new ValueEntityClientImpl(runtimeComponentClients.valueEntityClient, callMetadata, valueEntityId)

  override def forEventSourcedEntity(eventSourcedEntityId: String): EventSourcedEntityClient =
    EventSourcedEntityClientImpl(runtimeComponentClients.eventSourcedEntityClient, callMetadata, eventSourcedEntityId)

  override def forWorkflow(workflowId: String): WorkflowClient =
    WorkflowClientImpl(runtimeComponentClients.workFlowClient, callMetadata, workflowId)

  override def forView(): ViewClient = ViewClientImpl(runtimeComponentClients.viewClient, callMetadata)

}
